### PR TITLE
Fractal node

### PR DESCRIPTION
**Pull request by kalwalt**
We continue the discussion on this issue: https://github.com/nortikin/sverchok/issues/1177  
See the extra comments, in particulary on https://github.com/kalwalt/sverchok/commit/e81fcd5e05138b281a22d3d3786f9fddefc924e6

### CODE REVIEW

**Comment by zeffii** [comment]
do all of your fractal functions take the same number of arguments / have some function signature and do they all only return a float value?

**Comment by zeffii** [comment]
```
noise.fractal                        (position, H, lacunarity, octaves, noise_basis=noise.types.STDPERLIN)
noise.multi_fractal              (position, H, lacunarity, octaves, noise_basis=noise.types.STDPERLIN)
noise.hetero_terrain           (position, H, lacunarity, octaves, offset, noise_basis=noise.types.STDPERLIN)
noise.ridged_multi_fractal  (position, H, lacunarity, octaves, offset, gain, noise_basis=noise.types.STDPERLIN)
noise.hybrid_multi_fractal  (position, H, lacunarity, octaves, offset, gain, noise_basis=noise.types.STDPERLIN)
```

so, no infact hetero has an offset param,, and ridged/hybrid have offset+gain as extra parameters, but they all return a float, so no Vector output need and no output switching.

**Comment by zeffii** [comment]
i haven't tested this, but i suspect this is closer to what you want
```python
# ##### BEGIN GPL LICENSE BLOCK #####
#
#  This program is free software; you can redistribute it and/or
#  modify it under the terms of the GNU General Public License
#  as published by the Free Software Foundation; either version 2
#  of the License, or (at your option) any later version.
#
#  This program is distributed in the hope that it will be useful,
#  but WITHOUT ANY WARRANTY; without even the implied warranty of
#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#  GNU General Public License for more details.
#
#  You should have received a copy of the GNU General Public License
#  along with this program; if not, write to the Free Software Foundation,
#  Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
#
# ##### END GPL LICENSE BLOCK #####


import bpy
from bpy.props import EnumProperty, IntProperty, FloatProperty
from mathutils import noise

from sverchok.node_tree import SverchCustomTreeNode
from sverchok.data_structure import updateNode


def fractal(v, h_factor, lacunarity, octaves, offset, gain, _noise_type):
    return noise.fractal(verts, h_factor, lacunarity, octaves, _noise_type)
 
def multifractal(v, h_factor,  lacunarity, octaves, offset, gain, _noise_type):
    out = 0.0
    print("hey, i'am a multifractal!")
    return out

def hetero(v, h_factor,  lacunarity, octaves, offset, gain, _noise_type):
    out = 0.0
    print("hetero fractal works")
    return out

def ridged(v, h_factor,  lacunarity, octaves, offset, gain, _noise_type):
    out = 0.0
    print("ridged fractal!")
    return out

def hybrid(v, h_factor,  lacunarity, octaves, offset, gain, _noise_type):
    out = 0.0
    print("you got a hybrid fractal!")
    return out

fractal_f = {
    'FRACTAL': fractal, 
    'MULTI_FRACTAL': multifractal, 
    'HETERO_TERRAIN': hetero, 
    'RIDGED_MULTI_FRACTAL': ridged, 
    'HYBRID_MULTI_FRACTAL': hybrid
}

# noise nodes
# from http://www.blender.org/documentation/blender_python_api_current/mathutils.noise.html
noise_options = [
    ('BLENDER', 0),
    ('STDPERLIN', 1),
    ('NEWPERLIN', 2),
    ('VORONOI_F1', 3),
    ('VORONOI_F2', 4),
    ('VORONOI_F3', 5),
    ('VORONOI_F4', 6),
    ('VORONOI_F2F1', 7),
    ('VORONOI_CRACKLE', 8),
    ('CELLNOISE', 14)
]

fractal_options = [
    ('FRACTAL', 0),
    ('MULTI_FRACTAL', 1),
    ('HETERO_TERRAIN', 2),
    ('RIDGED_MULTI_FRACTAL', 3),
    ('HYBRID_MULTI_FRACTAL', 4),
]


noise_dict = {t[0]: t[1] for t in noise_options}
avail_noise = [(t[0], t[0].title(), t[0].title(), '', t[1]) for t in noise_options]

fractal_dict = {t[0]: t[1] for t in fractal_options}
avail_fractal = [(t[0], t[0].title(), t[0].title(), '', t[1]) for t in fractal_options]


class SvVectorFractal(bpy.types.Node, SverchCustomTreeNode):
    '''Vector Fractal node'''
    bl_idname = 'SvVectorFractal'
    bl_label = 'Vector Fractal'
    bl_icon = 'FORCE_TURBULENCE'

    noise_type = EnumProperty(
        items=avail_noise,
        default='STDPERLIN',
        description="Noise type",
        update=updateNode)

    fractal_type = EnumProperty(
        items=avail_fractal,
        default="FRACTAL",
        description="Fractal type",
        update=updateNode)

    h_factor = FloatProperty(default=0.5, name='H Factor', update=updateNode)
    lacunarity = FloatProperty(default=1.0, name='Lacunarity', update=updateNode)
    octaves = IntProperty(default=3, min=0, max=6, name='Octaves', update=updateNode)
    offset = FloatProperty(default=0.0, name='Offset', update=updateNode)
    gain = FloatProperty(default=0.5, name='Gain', update=updateNode)

    def sv_init(self, context):
        self.inputs.new('VerticesSocket', 'Vertices')
        self.inputs.new('StringsSocket', 'H Factor').prop_name = 'h_factor'
        self.inputs.new('StringsSocket', 'Lacunarity').prop_name = 'lacunarity'
        self.inputs.new('StringsSocket', 'Octaves').prop_name = 'octaves'
        self.inputs.new('StringsSocket', 'Offset').prop_name = 'offset'
        self.inputs.new('StringsSocket', 'Gain').prop_name = 'gain'        
        self.outputs.new('StringsSocket', 'Value')

    def draw_buttons(self, context, layout):
        layout.prop(self, 'fractal_type', text="Type")
        layout.prop(self, 'noise_type', text="Type")

    def process(self):
        inputs, outputs = self.inputs, self.outputs

        if not outputs[0].is_linked:
            return

        out = []
        verts = inputs['Vertices'].sv_get()
        h_factor = inputs['H Factor'].sv_get()[0][0]
        lacunarity = inputs['Lacunarity'].sv_get()[0][0]
        octaves = inputs['Octaves'].sv_get()[0][0]
        offset = inputs['Offset'].sv_get()[0][0]
        gain = inputs['Gain'].sv_get()[0][0]

        _noise_type = noise_dict[self.noise_type]
        fractal_function = fractal_f[self.fractal_type]

        for vlist in verts:
            out.append([fractal_function(v, h_factor, lacunarity, octaves, offset, gain, _noise_type) for v in vlist])

        outputs[0].sv_set(out)


def register():
    bpy.utils.register_class(SvVectorFractal)


def unregister():
    bpy.utils.unregister_class(SvVectorFractal)

```

**Comment by zeffii** [comment]
unfortunately we didn't add a way to "grey out / make inactive" a slider when the mode doesn't need that parameter. to do so isn't too hard, but we'd need to be certain that it doesn't break existing .blend / layouts is a little trickier.

**Comment by kalwalt** [comment]
@zeffii this is exactly what i thought to do, i will test it as i have a bit of time.

**Comment by zeffii** [comment]
try this
```python
# ##### BEGIN GPL LICENSE BLOCK #####
#
#  This program is free software; you can redistribute it and/or
#  modify it under the terms of the GNU General Public License
#  as published by the Free Software Foundation; either version 2
#  of the License, or (at your option) any later version.
#
#  This program is distributed in the hope that it will be useful,
#  but WITHOUT ANY WARRANTY; without even the implied warranty of
#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#  GNU General Public License for more details.
#
#  You should have received a copy of the GNU General Public License
#  along with this program; if not, write to the Free Software Foundation,
#  Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
#
# ##### END GPL LICENSE BLOCK #####


import bpy
from bpy.props import EnumProperty, IntProperty, FloatProperty
from mathutils import noise

from sverchok.node_tree import SverchCustomTreeNode
from sverchok.data_structure import updateNode


def fractal(_noise_type, v, h_factor, lacunarity, octaves, offset, gain):
    return noise.fractal(verts, h_factor, lacunarity, octaves, _noise_type)
 
def multifractal(_noise_type, v, h_factor, lacunarity, octaves, offset, gain):
    out = 0.0
    print("hey, i'am a multifractal!")
    return out

def hetero(_noise_type, v, h_factor, lacunarity, octaves, offset, gain):
    out = 0.0
    print("hetero fractal works")
    return out

def ridged(_noise_type, v, h_factor, lacunarity, octaves, offset, gain):
    out = 0.0
    print("ridged fractal!")
    return out

def hybrid(_noise_type, v, h_factor, lacunarity, octaves, offset, gain):
    out = 0.0
    print("you got a hybrid fractal!")
    return out

fractal_f = {
    'FRACTAL': fractal, 
    'MULTI_FRACTAL': multifractal, 
    'HETERO_TERRAIN': hetero, 
    'RIDGED_MULTI_FRACTAL': ridged, 
    'HYBRID_MULTI_FRACTAL': hybrid
}

# noise nodes
# from http://www.blender.org/documentation/blender_python_api_current/mathutils.noise.html
noise_options = [
    ('BLENDER', 0),
    ('STDPERLIN', 1),
    ('NEWPERLIN', 2),
    ('VORONOI_F1', 3),
    ('VORONOI_F2', 4),
    ('VORONOI_F3', 5),
    ('VORONOI_F4', 6),
    ('VORONOI_F2F1', 7),
    ('VORONOI_CRACKLE', 8),
    ('CELLNOISE', 14)
]

fractal_options = [
    ('FRACTAL', 0),
    ('MULTI_FRACTAL', 1),
    ('HETERO_TERRAIN', 2),
    ('RIDGED_MULTI_FRACTAL', 3),
    ('HYBRID_MULTI_FRACTAL', 4),
]


noise_dict = {t[0]: t[1] for t in noise_options}
avail_noise = [(t[0], t[0].title(), t[0].title(), '', t[1]) for t in noise_options]

fractal_dict = {t[0]: t[1] for t in fractal_options}
avail_fractal = [(t[0], t[0].title(), t[0].title(), '', t[1]) for t in fractal_options]


class SvVectorFractal(bpy.types.Node, SverchCustomTreeNode):
    '''Vector Fractal node'''
    bl_idname = 'SvVectorFractal'
    bl_label = 'Vector Fractal'
    bl_icon = 'FORCE_TURBULENCE'

    noise_type = EnumProperty(
        items=avail_noise,
        default='STDPERLIN',
        description="Noise type",
        update=updateNode)

    fractal_type = EnumProperty(
        items=avail_fractal,
        default="FRACTAL",
        description="Fractal type",
        update=updateNode)

    h_factor = FloatProperty(default=0.5, name='H Factor', update=updateNode)
    lacunarity = FloatProperty(default=1.0, name='Lacunarity', update=updateNode)
    octaves = IntProperty(default=3, min=0, max=6, name='Octaves', update=updateNode)
    offset = FloatProperty(default=0.0, name='Offset', update=updateNode)
    gain = FloatProperty(default=0.5, name='Gain', update=updateNode)

    def sv_init(self, context):
        self.inputs.new('VerticesSocket', 'Vertices')
        self.inputs.new('StringsSocket', 'H Factor').prop_name = 'h_factor'
        self.inputs.new('StringsSocket', 'Lacunarity').prop_name = 'lacunarity'
        self.inputs.new('StringsSocket', 'Octaves').prop_name = 'octaves'
        self.inputs.new('StringsSocket', 'Offset').prop_name = 'offset'
        self.inputs.new('StringsSocket', 'Gain').prop_name = 'gain'        
        self.outputs.new('StringsSocket', 'Value')

    def draw_buttons(self, context, layout):
        layout.prop(self, 'fractal_type', text="Type")
        layout.prop(self, 'noise_type', text="Type")

    def process(self):
        inputs, outputs = self.inputs, self.outputs

        if not outputs[0].is_linked:
            return

        out = []
        _noise_type = noise_dict[self.noise_type]
        fractal_function = fractal_f[self.fractal_type]

        verts = inputs['Vertices'].sv_get()

        m_h_factor = inputs['H Factor'].sv_get()[0]
        m_lacunarity = inputs['Lacunarity'].sv_get()[0]
        m_octaves = inputs['Octaves'].sv_get()[0]
        m_offset = inputs['Offset'].sv_get()[0]
        m_gain = inputs['Gain'].sv_get()[0]

        param_list = [m_h_factor, m_lacunarity, m_octaves, m_offset, m_gain]

        for idx, vlist in enumerate(verts):
            # lazy generation of full parameters.
            params = [(param[idx] if idx < len(param) else param[-1]) for param in param_list]
            out.append([fractal_function(_noise_type, v, *params) for v in vlist])

        outputs[0].sv_set(out)


def register():
    bpy.utils.register_class(SvVectorFractal)


def unregister():
    bpy.utils.unregister_class(SvVectorFractal)
```

**Comment by zeffii** [comment]
it's most convenient for me to test locally in my own branch..
https://github.com/nortikin/sverchok/pull/1182

**Comment by zeffii** [comment]
image

**Comment by zeffii** [comment]
Try this branch if you want
@kalwalt  for those functions  that don't accept Offset / Gain, maybe it makes sense to `value +offset *gain` the output  :)

**Comment by kalwalt** [comment]
@zeffii i'm looking into it, 

>for those functions that don't accept Offset / Gain, maybe it makes sense to value +offset *gain the output :)
Waht do you mean exactly? i wrote the 'fractal_function' with all the parameters because i hadn't a clear idea how to set the parameters for the different fractal functions.

**Comment by zeffii** [comment]
for example
```
def fractal(nbasis, verts, h_factor, lacunarity, octaves, offset, gain):
    return [noise.fractal(v, h_factor, lacunarity, octaves, nbasis) for v in verts]
```
a gain could be simple ampliciation (or some other parametrised function of current value)
```
def fractal(nbasis, verts, h_factor, lacunarity, octaves, offset, gain):
    return [(noise.fractal(v, h_factor, lacunarity, octaves, nbasis)*gain)+offset for v in verts]
```

**Comment by zeffii** [comment]
it's not clear from the docs what sensible defaults would be .. or min/max

**Comment by kalwalt** [comment]
Shall add more info on the descriptions? how long can be a description? should be fine add a link to an online/offline docs reference?

**Comment by kalwalt** [comment]
correct: the link to online/offline docs exist already...

**Comment by zeffii** [comment]
if you added docs, you need to `git add` them, and make sure to include the node in `vector_index.rst`

**Comment by ly29** [comment]
A bit mixed feelings about the graying out. Obviously the socket code is a complete mess already, mostly my fault.

I think it might better to remove and re add the socket when needed instead of graying it out.

**Comment by zeffii** [comment]
The reason i'm not in favour of removing the socket here is because switching in between modes would drop the links to those last params. The `<Layout>.enable` is used in many places in Blender - I find it useful to see parameters that might be available in other modes.. 

but do we need this feature ?  i don't know. I did part of that code out of curiosity.

**Comment by ly29** [comment]
We do drop such things in many other places like the math nodes but I do see your point.

**Comment by zeffii** [comment]
@kalwalt  also consider reusing the fractal_options list..

..snipped
```python
def hybrid(nbasis, verts, h_factor, lacunarity, octaves, offset, gain):
    return [noise.hybrid_multi_fractal(v, h_factor, lacunarity, octaves, offset, gain, nbasis) for v in verts]

# noise nodes
# from http://www.blender.org/documentation/blender_python_api_current/mathutils.noise.html
noise_options = [
    ('BLENDER', 0),
    ('STDPERLIN', 1),
    ('NEWPERLIN', 2),
    ('VORONOI_F1', 3),
    ('VORONOI_F2', 4),
    ('VORONOI_F3', 5),
    ('VORONOI_F4', 6),
    ('VORONOI_F2F1', 7),
    ('VORONOI_CRACKLE', 8),
    ('CELLNOISE', 14)
]

fractal_options = [
    ('FRACTAL', 0, (0, 0), fractal),
    ('MULTI_FRACTAL', 1, (0, 0), multifractal),
    ('HETERO_TERRAIN', 2, (1, 0), hetero),
    ('RIDGED_MULTI_FRACTAL', 3, (1, 1), ridged),
    ('HYBRID_MULTI_FRACTAL', 4, (1, 1), hybrid),
]


noise_dict = {t[0]: t[1] for t in noise_options}
avail_noise = [(t[0], t[0].title(), t[0].title(), '', t[1]) for t in noise_options]

fractal_f = {t[0]: t[3] for t in fractal_options}
fractal_dict = {t[0]: t[1] for t in fractal_options}
avail_fractal = [(t[0], t[0].title(), t[0].title(), '', t[1]) for t in fractal_options]
props_enabled = {t[0]: t[2] for t in fractal_options}


class SvVectorFractal(bpy.types.Node, SverchCustomTreeNode):
    '''Vector Fractal node'''
    bl_idname = 'SvVectorFractal'

```

**Comment by zeffii** [comment]
also
```python
noise_dict = {t[0]: t[1] for t in noise_options}
avail_noise = [(t[0], t[0].title(), t[0].title(), '', t[1]) for t in noise_options]

fractal_f = {t[0]: t[3] for t in fractal_options}
props_enabled = {t[0]: t[2] for t in fractal_options}
avail_fractal = [(t[0], t[0].title(), t[0].title(), '', t[1]) for t in fractal_options]
```
vs
```python
def dict_from(options, idx1, idx2):
    return {t[idx1]: t[idx2] for t in noise_options}

def enum_from(options):
    return [(t[0], t[0].title(), t[0].title(), '', t[1]) for t in options]

noise_dict = dict_from(noise_options, 0, 1)
avail_noise = enum_from(noise_options)

fractal_f = dict_from(fractal_options, 0, 3)
props_enabled = dict_from(fractal_options, 0, 2)
avail_fractal = enum_from(fractal_options)
```

**Comment by zeffii** [comment]
https://gist.github.com/zeffii/24e5375abf0ebef8f8813b592d1dc177

**Comment by kalwalt** [comment]
ok, i will do these changes, Iìm looking into it.

**Comment by kalwalt** [comment]
I'm getting this error now:

> File "/home/walter/blender-2.78a-linux-glibc211-x86_64/2.78/scripts/addons/sverchok-master/core/update_system.py", line 322, in do_update_general
    node.process()
  File "/home/walter/blender-2.78a-linux-glibc211-x86_64/2.78/scripts/addons/sverchok-master/nodes/vector/vector_rewire.py", line 117, in process
    series_vec.append(list(coords))
  File "/home/walter/blender-2.78a-linux-glibc211-x86_64/2.78/scripts/addons/sverchok-master/nodes/vector/vector_rewire.py", line 116, in <genexpr>
    coords = ([v[0], v[1], next(yield_value)] for v in obj)
Node Vector Rewire had exception 'float' object is not subscriptable

i have a vector rewire node plugged into the output of a vector fractal node:
screen2017-01-26-18-15- vectorfractalnode

**Comment by zeffii** [comment]
you are wiring a StringsSocket (floats...) to a VerticesSocket..

**Comment by kalwalt** [comment]
But i selected the scalar flag, this is not possible anymore?

**Comment by kalwalt** [comment]
no, my mistake wrong connections, works fine, sorry!

**Comment by zeffii** [comment]
that's the same as doing 
image

scalar on Vector Rewire refers to the replament of a component with some list of numbers.. see

image

**Comment by zeffii** [comment]
@kalwalt   the mistake you made is also a prime example of why we normally remove a socket when a mode doesn't require the socket. The question is really would it help if I also used the `layout.enable` trick there or if I added code that would remove/add that socket whenever the mode dictates.

**Comment by zeffii** [comment]
here's how it might look if you remove the sockets instead of greying them out.
https://gist.github.com/zeffii/a60dc790273747ec07defe5ce282201c

**Comment by kalwalt** [comment]
@zeffii i will look at tomorrow, thank's!

File: node_tree.py [hunk 1]
```diff
@@ -231,6 +231,8 @@ class StringsSocket(NodeSocket, SvSocketCommon):
 
     prop_type = StringProperty(default='')
     prop_index = IntProperty()
+    prop_enabled = BoolProperty(default=True)
```
Comment (zeffii) [inline node_tree.py]: we dont need this now :)

File: node_tree.py [hunk 2]
```diff
@@ -267,7 +269,10 @@ def draw(self, context, layout, node, text):
 
         if not self.is_output and not self.is_linked:
             if self.prop_name and not self.prop_type:
+
+                layout.enabled = self.prop_enabled
```
Comment (zeffii) [inline node_tree.py]: this can also go

**Comment by zeffii** [comment]
also undo the changes made to node_tree.py  (i've commented the file)

**Comment by zeffii** [comment]
after than I don't really have any issues accepting  the PR

**Comment by kalwalt** [comment]
Fixed! Should be OK now...

**Comment by kalwalt** [comment]
I'm also working on a turbulence node, i will create a separate branch and will start a new issue.

**Comment by zeffii** [comment]
congratulations on first Node ! 

I think it might need one or two small changes to work with iosjon (our import / export system) - but that's easier to implement as part of master

**Comment by ly29** [comment]
Yeah nice work indeed

**Comment by kalwalt** [comment]
> congratulations on first Node !
yeah, lol! really, thanks to all indeed, i learnt a lot! :smile:

**Comment by zeffii** [comment]
btw, just tested with the IO, works fine
https://gist.github.com/3f59a055c578ba21cb16dd6372d70454

**Pull request merged**